aiter/ops/triton/utils/mha_onekernel_utils.py

```diff
@@ -0,0 +1,1157 @@
+                )
+            else:
+                q = generate_bhsd_tensor(
+                    BATCH,
+                    HQ,
+                    N_CTX_Q,
+                    D_HEAD,
+                    dtype=dtype,
+                    device=device,
+                    DEBUG_INPUT=DEBUG_INPUT,
+                )
+                k = generate_bhsd_tensor(
+                    BATCH,
+                    HK,
+                    N_CTX_K,
+                    D_HEAD,
+                    dtype=dtype,
+                    device=device,
+                    DEBUG_INPUT=DEBUG_INPUT,
+                )
+                v = generate_bhsd_tensor(
+                    BATCH,
+                    HK,
+                    N_CTX_K,
+                    D_HEAD,
+                    dtype=dtype,
+                    device=device,
+                    DEBUG_INPUT=DEBUG_INPUT,
+                )
+                do = torch.ones_like(q) if DEBUG_INPUT else torch.randn_like(q)
+
+        # setup metadata
+        if DEBUG_INPUT:
+            sm_scale = 1
+        else:
+            sm_scale = D_HEAD**-0.5
+        metadata = MetaData(sm_scale=sm_scale)
+        metadata.max_seqlens_q = N_CTX_Q
+        metadata.max_seqlens_k = N_CTX_K
+        metadata.layout = layout
+        metadata.need_causal(CAUSAL)
+        metadata.need_dropout(DROPOUT_P)
+    else:
+        raise ValueError(f"Unknown layout: {layout}")
+
+    # deal with packing
+    if packing is None:
+        if is_fp8_dtype:
+            return (
+                (q, descale_q),
+                (k, descale_k),
+                (v, descale_v),
+                (do, descale_do),
+                metadata,
+            )
+        else:
+            return q, k, v, do, metadata
+    elif packing == "kv":
+        # pack k and v
+        if layout in ["bhsd", "thd"]:
+            kv = torch.stack([k, v], dim=1)
+        elif layout == "bshd":
+            kv = torch.stack([k, v], dim=2)
+        else:
+            raise ValueError(f"Unknown layout: {layout}")
+
+        if is_fp8_dtype:
+            raise ValueError("FP8 not supported kv packing yet")
+        else:
+            return q, kv, do, metadata
+    elif packing == "qkv":
+        # qkv packing - requires same sequence length for q and k
+        assert (
+            N_CTX_Q == N_CTX_K
+        ), "For QKV packing, Q and K must have same sequence length"
+        assert HQ == HK, "For QKV packing, Q and K must have same number of heads"
+
+        # pack q, k, and v
+        if layout in ["bhsd", "thd"]:
+            qkv = torch.stack([q, k, v], dim=1)
+        elif layout == "bshd":
+            qkv = torch.stack([q, k, v], dim=2)
+        else:
+            raise ValueError(f"Unknown layout: {layout}")
+
+        if is_fp8_dtype:
+            raise ValueError("FP8 not supported qkv packing yet")
+        else:
+            return qkv, do, metadata
+    else:
+        assert False, f"Unsupported packing mode: {packing}"
```

Comment:
These input helpers should move to op_tests. Also, we already have some for mha, so please make sure they are not repeated